app/src/main/java/com/example/testapp/ui/login/LoginViewModel.kt

```diff
@@ -1,11 +1,96 @@
 package com.example.testapp.ui.login
 
-
+import android.app.Application
+import android.view.View
+import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.example.testapp.data.Resource.Data
+
+import com.example.testapp.data.Resource.Error
+import com.example.testapp.data.LoginRequest
+
+import com.example.testapp.repository.LoginRepository
+
+
+/**
+ * ViewModel Class for Login View
+ *
+ * @property userRepository
+ */
+ class LoginViewModel(private val loginRepository: LoginRepository, application: Application) :
+    ViewModel() {
+
+    // Variable to check if response is successful
+    val responseStatus = MutableLiveData<Any>()
+
+    // Variable to check if response is successful
+    val responseStatusVerificationEmailPost = MutableLiveData<Boolean>()
+
+    // Variable for Login Form data
+    var login: LoginForm? = null
+
+    // Email On focus listener to handle validation
+    var emailOnFocusChangeListener: View.OnFocusChangeListener? = null
+
+    // Password on focus listener to handle validation
+    var passwordOnFocusChangeListener: View.OnFocusChangeListener? = null
+
+    // LiveData for observing change in loginField i.e if all validations are passed value is set and observed
+    val loginFields: MutableLiveData<LoginForm.LoginFields>?
+        get() = login?.loginFields
+
+    init {
+        login = LoginForm()
+
+    }
+
+    /**
+     * Attempt to authenticate user
+     */
+    suspend fun authenticateUser() {
+
+        val email = loginFields?.value?.email
```

Comment:
Please try using scoped functions here to avoid repetitive code: You can check about scoped functions here: https://blog.mindorks.com/using-scoped-functions-in-kotlin-let-run-with-also-apply